w3-aggregation.md

```diff
@@ -0,0 +1,337 @@
+
+### `aggregate/offer`
+
+A Storefront principal can invoke a capabilty to submit an aggregate ready for offers.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/offer",
+    "nb": {
+      "offer": {
+        "link": "bagy...aggregate",
+        "size": 110101,
+        "commP": "commP...",
+        "src": ["https://w3s.link/ipfs/bagy...aggregate"]
+      }
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+This capability is invoked to submit a request to a broker service when an aggregate is ready for a Filecoin deal. The `nb.offer.link` is the CAR CID that refers to a "Ferry" aggregate, a collection of dag-cbor blocks with format:
+
+```json
+{
+  "link": "bag...",
+  "size": 110101,
+  "commP": "commP...",
+  "src": ["https://.../bag(...).car"]
+}
+```
+
+A receipt will be generated to acknowledge the received request. This receipt MUST contain an [effect](https://github.com/ucan-wg/invocation/#7-effect) with a subsequent task (`.fx.join` field) that is run when submitted aggregate is processed and either succeeds (implying that aggregate was accepted and deals will be arranged) or fail (with `error` describing a problem with an aggregate)
+
+```json
+{
+  "ran": "bafy...invocation",
+  "out": {
+    "ok": {
+      "status": "queued"
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...dequeue" }
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+Open questions:
+
+- can we get `commP` of the aggregate with `commP` of every CAR that is part of it?
+
+### `aggregate/get`
+
+A Storefront principal can invoke a capability to get state of an accepted aggregate.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/get",
+    "nb": {
+      "commP": "commP...",
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+### `offer/review`
+
+When a broker receives an `aggregate/offer` invocation from a Storefront Principal, an [Effect](https://github.com/ucan-wg/invocation/#7-effect) for this submission is created with join task to be performed asynchronously.
+
+```json
+{
+  "iss": "did:web:spade.storage",
+  "aud": "did:web:web3.storage",
+  "att": [{
+    "with": "did:web:spade.storage",
+    "can": "offer/review",
+    "nb": {
+      "commP": "commP",
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+Once this invocation is executed, a receipt is generated with the status of the task updated. Accepted aggregate receipt will provide aggregate status info:
+
+```json
+{
+  "ran": "bafy...review",
+  "out": {
+    "ok": {
+       "status": "accepted",
+       "link": "bafy...aggregate"
```

Comment:
Why do we need this field?

Reply:
Yeah for convenience. If we don't have pointer to the aggregate you will not know for which aggregate this receipt is. You would have to first resolve invocation block from `ran` field, and find out the aggregate from it's `nb.link`. 

All of this can be made even more complicated if you don't have necessary blocks locally. Including the link here means you'll know all from the receipt without having to do any resolution.

Reply:
I mentioned it elsewhere, but inline here is probably going to make more sense.

In terms of capabilities it does seem wrong that broker is able to mutate `did:web:web3.storage` resource without been delegated capability to do so. And that is kind of what happens as a side effect of `offer/arrange`.

> Although perhaps we could make an argument that by delegating `aggregate/offer` you do delegate ability to also update it's state.

I think it would be more cleaner if instead `offer/arrange` produced a receipt which in success case would include delegation to `offer/*` capabilities to the storefront which would allow it then to query for the state using `offer/get` capability. With such arrangement it would make sense that spade can update own state and that storefront will be able to query it because it was delegated capabilities to do so.

As an added benefit it would also make it impossible to query for the status until aggregate was accepted, because storefront simple will not have capability to do so.

The downside of going this route is that storefront would need to maintain delegations for the accepted aggregates to query the state. However we could probably find a way around this by adding additional `nb.storefront` field into `offer/*` capabilities so that one delegation could cover all of them.